### PR TITLE
ROX-14078: fix CA setup test

### PR DESCRIPTION
**Pull request by janisz**
### Description

This PR changes the CA setup test so it's not relying on external services to validate if additional CA is setup correctly. Previously we have failures due to badssl.com was down. Now we don't even call it. 

- [x] CHANGELOG update is not needed
- [x] Documentation is not needed
### Testing
- [x] inspected CI results
#### Automated testing
- [x] modified existing tests
- [x] contributed **no automated tests**
#### How I validated my change
CI

### CODE REVIEW

**Review by porridge — COMMENTED** [review]
It looks like I'm missing a lot of background on this test.
And I'm utterly confused about what this `ca_setup_test.go` is supposed to test.

If we're reducing it to just cert parsing + Verify, then why not make it simply a unit test? Why do we need to call this development service?

Or perhaps this URLHasValidCert has actual production uses? In that case we're not really testing the production code path (which AFAICT is not providing a cert inline).

Perhaps @SimonBaeumer can shed some light on this area?

File: tests/ca_setup_test.go [hunk 1]
```diff
@@ -46,32 +46,106 @@ func TestCASetup(t *testing.T) {
...
+OXy+l+8L1gtY54x/v7NMHfO3kICmNnwUW+wHLQI+G1tjWxWrAPofOxkt3+IjEBEH
+fnJ/4r+3ABuYLyw/zoWaJ4wQIghBK4o+gk783SHGVnRwpDTysUCeK1iiWQ8dSO/r
+ET7BSp68ZVVtxqPv1dSWzfGuJ/ekVxQ8lEEFeouhN0fX9X3c+s5vMaKwjOrMEpsi
+8TRwz311SotoKQwe6Zaoz7ASH1wq7mcvf71z81oBIgxw+s1F73hczg36TuHvzmWf
+RwxPuzZEaFZcVlmtqoq8
+-----END CERTIFICATE-----`,
 			expectedResp: central.URLHasValidCertResponse_CERT_SIGNING_AUTHORITY_VALID_BUT_OTHER_ERROR,
 		},
-		{
-			url:          "https://google.com",
```
Comment (porridge) [inline tests/ca_setup_test.go:68]: Why drop these?

This should probably change to https://connectivitycheck.gstatic.com/

And the other to a non-existent subdomain of a domain that we do control.

Comment (janisz) [reply tests/ca_setup_test.go:68]: Good point `test.invalid` should be OK
https://www.rfc-editor.org/rfc/rfc2606.html
https://www.rfc-editor.org/rfc/rfc6761.html

**Comment by janisz** [comment]
> If we're reducing it to just cert parsing + Verify, then why not make it simply a unit test? Why do we need to call this development service?
I agree that is not trivial to understand. If I read code correctly this test purpose is to validate if additional CA is setup correctly that's why we need it as e2e test because we want to pass certificate from installation step all the way to the `SystemCertPool`. Previously we make a network call with default HTTP client to verify if cert was added correctly. This lead to a problem that we call a site that is not reliable (not meant to use in CI). To solve this issue I've change the code to not make a call but instead verify provided cert against system cert pool. There is a subtle difference between e2e and unit test. They look similar but there is one case that differs because in e2e we test additional CA.

https://github.com/stackrox/stackrox/blob/0d2691d8e4b11a802bae81661f5110e1da01376a/tests/e2e/run.sh#L102

> Or perhaps this URLHasValidCert has actual production uses? In that case we're not really testing the production code path (which AFAICT is not providing a cert inline).
No it has not, this service it's not present in release build

File: central/development/service/service_impl.go [hunk 2]
```diff
@@ -108,15 +110,22 @@ func (s *serviceImpl) ReconciliationStatsByCluster(context.Context, *central.Emp
...
 func (s *serviceImpl) URLHasValidCert(_ context.Context, req *central.URLHasValidCertRequest) (*central.URLHasValidCertResponse, error) {
-	if !strings.HasPrefix(req.GetUrl(), "https://") {
-		return nil, errors.Wrapf(errox.InvalidArgs, "url %q must start with https", req.GetUrl())
+	u, err := url.Parse(req.GetUrl())
+	if err != nil {
+		return nil, errors.Wrapf(errox.InvalidArgs, "invalid url %s", err.Error())
+	}
+
+	if req.CertPEM == "" {
+		_, err = s.client.Get(req.GetUrl())
```
Comment (porridge) [inline central/development/service/service_impl.go:119]: Can you please add a comment here how `client.Get` and `verifyProvidedCert` are in fact exercising the same validation logic and **this** is what this is all about, and how this logic is affected by the env variable. TBH it's a mystery to me how `client.Get` gets configured to respect that install-time variable, it looks like a generic 3rd-party library!

**Review by porridge — COMMENTED** [review]
Thanks for the explanation! I knew about this variable because that's also one of the causes for https://issues.redhat.com/browse/ROX-24688 but didn't catch the comment ` This should succeed because, even though it's a bad cert, we have configured Central to trust its root`. We should make this more obvious for posterity!

If we're now saving these certs in the repo, then to keep it robust from time passing PoV, we should make sure these hardcoded certs are valid for at least 50 years. We can create our own ones, and then save the creation commands in comments.

File: proto/internalapi/central/development_service.proto [hunk 3]
```diff
@@ -11,6 +11,7 @@ option java_package = "io.stackrox.proto.internalapi.central";
 
 message URLHasValidCertRequest {
```
[the remark this answers is not in the document]
Comment (janisz) [reply proto/internalapi/central/development_service.proto:12]: The host is part of the cert verification process. I think that's fine a it's internal service.

**Comment by janisz** [comment]
> If we're now saving these certs in the repo, then to keep it robust from time passing PoV, we should make sure these hardcoded certs are valid for at least 50 years. We can create our own ones, and then save the creation commands in comments.

That's good idea. I think we already have a script somewhere in the code as it failed some time ago.

**Comment by janisz** [comment]
Certs expires on May 17, 2026 except the expired one (that expires on: Apr 13, 2015) so I think we are fine :) 
But I can generate new ones

**Comment by porridge** [comment]
Please do generate new ones. Otherwise this PR changes a flaky test to a time bomb 😆

File: central/development/service/service_impl_test.go [hunk 4]
```diff
@@ -136,3 +139,127 @@ func (s *developmentServiceAccessControlTestSuite) TestDevelopmentServiceRandomB
...
+	cases := []struct {
+		url               string
+		certPEM           string
+		expectedResp      central.URLHasValidCertResponse_URLResult
+		additionalMessage string
+	}{
+		{
+			url: "https://untrusted-root.badssl.com",
+			certPEM: `
+-----BEGIN CERTIFICATE-----
```
Comment (SimonBaeumer) [inline central/development/service/service_impl_test.go:153]: Can you extend the validity of the certificates to unlimited? They are expiring in 2026, I've run into this issue in the `TLSChallengeTest` multiple times and regretted it.

**Review by SimonBaeumer — APPROVED** [review]
LGTM, only have a look at the cert expiry date.

**Comment by porridge** [comment]
@janisz setup is failing with
```
Trusted CA file /go/src/github.com/stackrox/stackrox/tests/bad-ca/root.pem not found
```

**Review by porridge — APPROVED** [review]
❤️

File: central/development/service/service_impl_test.go (hunk 4 above)
Comment (janisz) [reply central/development/service/service_impl_test.go]: 50 years should be fine

**Pull request merged**